evaluate.py

```diff
@@ -0,0 +1,115 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+
+from pyro.contrib.forecast.evaluate import backtest, logger
+from pyro.ops.stats import crps_empirical, quantile
+
+from util import M5Data
+
+
+@torch.no_grad()
+def eval_mae(pred, truth):
+    """
+    Like pyro.contrib.forecast.eval_mae but does not average over batch dimensions.
+    """
+    pred = pred.median(0).values
+    return (pred - truth).abs().reshape(truth.shape[:-2] + (-1,)).mean(-1)
+
+
+@torch.no_grad()
+def eval_rmse(pred, truth):
+    """
+    Like pyro.contrib.forecast.eval_rmse but does not average over batch dimensions.
+    """
+    pred = pred.mean(0)
+    error = pred - truth
+    return (error * error).reshape(truth.shape[:-2] + (-1,)).mean(-1).sqrt()
+
+
+@torch.no_grad()
+def eval_crps(pred, truth):
+    """
+    Like pyro.contrib.forecast.eval_crps but does not average over batch dimensions.
+    """
+    return crps_empirical(pred, truth).reshape(truth.shape[:-2] + (-1,)).mean(-1)
+
+
+@torch.no_grad()
+def eval_pl(pred, truth):
+    """
+    Computes pinball loss over 9 quantiles 0.005, 0.165, 0.25, 0.5, 0.75, 0.835, 0.975, 0.995.
+    """
+    us = torch.tensor(M5Data.quantiles, dtype=pred.dtype, device=pred.device)
+    pred = quantile(pred, probs=us, dim=0)  # 9 x batch_shape x duration x D
+    error = pred - truth.unsqueeze(0)
+    us = us.reshape((-1,) + (1,) * (pred.dim() - 1)).expand(pred.shape)
+    error = (torch.where(error <= 0, us, 1 - us) * error).mean(0)  # mean accross all quantiles
+    return error.reshape(truth.shape[:-2] + (-1,)).mean(-1)
+
+
+DEFAULT_METRICS = {
+    "mae": eval_mae,
+    "rmse": eval_rmse,
+    "crps": eval_crps,
+    "pl": eval_pl,
+}
+
+
+def _get_metric_scale(metric, train_data):
+    duration = train_data.shape[-2]
+    lag1 = train_data - torch.nn.functional.pad(train_data[..., :-1, :], (0, 0, 1, 0))
+    # find active time: to drop the leading 0s
+    active_time = (train_data.sum(-1, keepdims=True).cumsum(-2) != 0).sum(-2, keepdims=True)
+    start_value = train_data.gather(
+        -2, (duration - active_time).expand(active_time.shape[:-1] + train_data.shape[-1:]))
+    norm = 2 if metric == "rmse" else 1
+    lag1_norm = lag1.abs().pow(norm).sum(-2) - start_value.squeeze(-2).abs().pow(norm)
```

Comment:
I don't quite understand the `start_value` bit; could it be avoided if you instead defined
```py
lag1 = tran_data[..., 1:, :] - train_data[..., :-1, :]
```

Reply:
This is a bit subtle: assume x = train_data = (0, 0, 1, 3, 6); the active series will be (1, 3, 6) and active lag1 is (2, 3). If we use lag1 = x[1:] - x[:-1], then lag1 = (0, 1, 2, 3). So we need to remove `1`, which is the `start_value`, to get active lag1.